File: 2020/12.py
ls = [[l[0], int(l[1:])] for l in open("../inputs/2020/12.txt", "r").readlines()]

ms = { "N": [1, 0], "S": [-1, 0], "E": [0, 1], "W": [0, -1] }
ds = { 0: "N", 90: "E", 180: "S", 270: "W" }
rs = { "R": [-1, 1], "L": [1, -1] }

p1,d,p2,wp = [0, 0],90,[0, 0],[1, 10]
for l in ls:
    if l[0] in ms:
        p1 = [p1[i] + ms[l[0]][i] * l[1] for i in range(2)]
        wp = [wp[i] + ms[l[0]][i] * l[1] for i in range(2)]
    if l[0] == "F":
        p1 = [p1[i] + ms[ds[d]][i] * l[1] for i in range(2)]
        p2 = [p2[i] + wp[i] * l[1] for i in range(2)]
    if l[0] in rs:
        d = (d + rs[l[0]][1] * l[1]) % 360
        for i in range(0, l[1]//90):
            wp = [rs[l[0]][0] * wp[1], rs[l[0]][1] * wp[0]]

print("Part 1:", abs(p1[0])+abs(p1[1]))
print("Part 2:", abs(p2[0])+abs(p2[1]))

File: 2023/cpp/12.cpp
#include <iostream>
#include <sstream>
#include <vector>
#include <tuple>
#include <map>

typedef long long int ll;

ll a, b;
std::string s, nums, n;
std::vector<int> ns;
std::map<std::tuple<int, int, int>, ll> cache;

ll fs(int i, int hc, int c) {
    std::tuple k = std::make_tuple(i, hc, c);
    if (cache.find(k) != cache.end())
        return cache[k];

    ll a = 0;
    if (c == ns.size())
        a = (s.find('#', i) == std::string::npos)? 1 : 0;
    else if (i == s.size())
        a = (ns.size() == (hc == ns[c] ? c + 1 : 0)) ? 1 : 0;
    else if (s[i] == '#')
        a = (hc < ns[c]) ? fs(i + 1, hc + 1, c) : 0;
    else {
        a = hc == ns[c] ? fs(i + 1, 0, c + 1) : (hc == 0 ? fs(i + 1, 0, c) : 0);
        a += (s[i] == '.') ? 0 : fs(i + 1, hc + 1, c);
    }
    cache[k] = a;
    return a;
}

int main() {
    freopen("../../inputs/2023/12.txt", "r", stdin);
    while (std::cin >> s >> nums) {
        std::istringstream nss(nums);
        ns.clear();
        cache.clear();
        while (std::getline(nss, n, ',')) {
            ns.push_back(std::stoi(n));
        }
        a += fs(0, 0, 0);

        cache.clear();
        int nsl = ns.size();
        std::string sc = s;
        for (int i = 0; i < 4; i++) {
            s += "?" + sc;
            for (int j = 0; j < nsl; j++)
                ns.push_back(ns[j]);
        }
        b += fs(0, 0, 0);
    }

    std::cout << "Day 12: Hot Springs" << std::endl;
    std::cout << "  Part 1: " << a << std::endl;
    std::cout << "  Part 2: " << b << std::endl;
}
